python/nutpie/compile_pymc.py

```diff
@@ -258,22 +258,145 @@ def compile_pymc_model(model: "pm.Model", **kwargs) -> CompiledPyMCModel:
+            "Please install it with something like "
+            "'mamba install -c conda-forge jax' "
+            "and restart your kernel in case you are in an interactive session."
+        )
+    import jax
+
+    if gradient_backend is None:
+        gradient_backend = "pytensor"
+    elif gradient_backend not in ["jax", "pytensor"]:
+        raise ValueError(f"Unknown gradient backend: {gradient_backend}")
+
+    (
+        n_dim,
+        _,
+        logp_fn_pt,
+        expand_fn_pt,
+        shape_info,
+    ) = _make_functions(
+        model,
+        mode="JAX",
+        compute_grad=gradient_backend == "pytensor",
+        join_expanded=False,
+    )
+
+    logp_fn = logp_fn_pt.vm.jit_fn
+    expand_fn = expand_fn_pt.vm.jit_fn
+
+    if gradient_backend == "jax":
+        orig_logp_fn = logp_fn._fun
+
+        @jax.jit
+        def logp_fn_jax_grad(x, **shared):
+            return jax.value_and_grad(lambda x: orig_logp_fn(x, **shared)[0])(x)
+
+        logp_fn = logp_fn_jax_grad
+
+    shared_data = {}
+    shared_vars = {}
+    seen = set()
+    for val in [*logp_fn_pt.get_shared(), *expand_fn_pt.get_shared()]:
+        if val.name in shared_data and val not in seen:
+            raise ValueError(f"Shared variables must have unique names: {val.name}")
+        shared_data[val.name] = jax.numpy.asarray(val.get_value().copy())
+        shared_vars[val.name] = val
+        seen.add(val)
+
+    logp_shared_names = [var.name for var in logp_fn_pt.get_shared()]
+    expand_shared_names = [var.name for var in expand_fn_pt.get_shared()]
+
+    def make_logp_func():
+        def logp(x, **shared):
+            logp, grad = logp_fn(
+                x, **{name: shared[name] for name in logp_shared_names}
+            )
+            return float(logp), np.asarray(grad, dtype="float64", order="C")
+
+        return logp
+
+    names, slices, shapes = shape_info
+    dtypes = [np.float64] * len(names)
+
+    def make_expand_func(seed1, seed2, chain):
+        # TODO handle seeds
+        def expand(x, **shared):
+            values = expand_fn(
+                x, **{name: shared[name] for name in expand_shared_names}
+            )
+            return {
+                name: np.asarray(val, order="C", dtype=dtype).ravel()
+                for name, val, dtype in zip(names, values, dtypes, strict=True)
+            }
+
+        return expand
+
+    dims, coords = _prepare_dims_and_coords(model, shape_info)
+
+    return from_pyfunc(
+        n_dim,
+        make_logp_func,
+        make_expand_func,
+        dtypes,
+        shapes,
+        names,
         shared_data=shared_data,
-        user_data=user_data,
-        n_expanded=n_expanded,
-        shape_info=shape_info,
-        logp_func=logp_fn_pt,
-        expand_func=expand_fn_pt,
+        dims=dims,
+        coords=coords,
     )
 
 
+def compile_pymc_model(
+    model: "pm.Model", *, backend="numba", gradient_backend=None, **kwargs
```

Comment:
String arguments is a place where I really like to have literal type hints
```suggestion
    model: "pm.Model", *, backend: Literal["numba", "jax"] ="numba", gradient_backend=None, **kwargs
```